boefjes/boefjes/katalogus/local_repository.py

```diff
@@ -141,6 +142,7 @@ def _boefje_to_plugin(boefje: BoefjeResource) -> Boefje:
         def_file = boefje.path / "boefje.json"
         def_obj = json.loads(def_file.read_text())
         def_obj["repository_id"] = RESERVED_LOCAL_ID
+        def_obj["runnable_hash"] = get_runnable_hash(boefje.path)
```

Comment:
I was also thinking that the `BoefjeResource` is loaded every time `resolve_boefjes()` and `by_id` are called, which is every time a job is created. Meaning we iterate repeatedly over the filesystem, create importable packages of, and calculate hashes for, all Boefjes in the katalogus. Not the most efficient flow if we're only interested in running a single Boefje. Candidate for future refactoring?

Reply:
Great observation. Modern OSs and filesystem drivers actually do quite a good job at (partially) caching these kinds of situations, but there's still a lot of performance gain if we cache this ourselves. Listing directory content isn't a big deal, but extensively reading out files and calculating hashes is quite expensive. Definitely a candidate for future refactoring. But we need to select a good cache key which usually is a combination of the file name and modification time.

Reply:
Only calculating the hash before a Boefje job is ran makes more sense to me. Can we change the API to make this possible?

Reply:
Rescanning the boefjes is quite nice in terms of quick feedback during development, so let's profile to verify the impact before creating some sort of caching layer.

Reply:
Container images already have hashes so no need to calculate any hashes for that. What we do here is with every call to those methods we calculate the hashes of files shipped with OpenKAT. It would just be a waste of resources, we could even calculate the hashes once in CI for everybody or just record the OpenKAT version / commit id that was used.

And if a malicious user can change the code of an OpenKAT install I don't see how generating hashing is going to help anything because the malicious user can also change the code that does the hashing.

Reply:
We are currently not creating those containers, nor hashing them. 
This is a step in the right direction to do so for plain python code. Yes, there are ways around this, but doing things reasonably well, is better than not doing them at all.
The usecase for an auditor is to spot changes in code since it was ran. This hash will end up in the timestamped remote signed hash, and as such gives the auditor a reasonable guess that the code that was used to generate a result is the code he is seeing in front of them (and can hash to produce the same result). Since hashing the pyc files produces possibly changing hashes depending on the platform, runtime or time of day, those should not be part of that proof. 

If someone is willing to change their OpenKAT install, mock the hashing code with something they can influence in order to produce timestamped proofs of data that they are presenting to their auditor, there's not much we can do expect run a trusted OpenKAT build on a trusted computer. And guess what? Those users who indent to fool their auditors wont use that system, and instead will opt just to not run OpenKAT and not have and proofs for their auditing or compliance statements.